### ISSUE TITLE
Observing "OutOfDirectMemoryError" exception in Bookkeeper

### ISSUE THREAD

**Issue by vedanthh** [pravega/pravega#3987]
**Observing `OutOfDirectMemoryError` exception -- ERROR - [bookie-io-1-2:BookieRequestHandler@70] - Unhandled exception occurred in I/O thread or handler
io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 1073741824, max: 1073741824) while running moderate IO workload with ( 12 readers and 7 writers )**

Environment details: PKS / K8 with medium cluster:
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
```
Pravega version : 0.5.0-2284.8c6c176
Zookeeper Operator : pravega/zookeeper-operator:0.2.2
Pravega Operator: pravega/pravega-operator:0.3.2
```
**Snip of error:**
[Bookkeeper_log.zip](https://github.com/pravega/pravega/files/3346101/Bookkeeper_log.zip)

```
2019-06-28 19:02:05,308 - ERROR - [bookie-io-1-2:BookieRequestHandler@70] - Unhandled exception occurred in I/O thread or handler
io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 1073741824, max: 1073741824)
        at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:640)
        at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:594)
        at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:764)
        at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:740)
        at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:244)
        at io.netty.buffer.PoolArena.allocate(PoolArena.java:226)
        at io.netty.buffer.PoolArena.allocate(PoolArena.java:146)
        at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:324)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:185)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:176)
        at io.netty.buffer.AbstractByteBufAllocator.ioBuffer(AbstractByteBufAllocator.java:137)
        at io.netty.channel.DefaultMaxMessagesRecvByteBufAllocator$MaxMessageHandle.allocate(DefaultMaxMessagesRecvByteBufAllocator.java:114)
        at io.netty.channel.epoll.EpollRecvByteAllocatorHandle.allocate(EpollRecvByteAllocatorHandle.java:71)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:791)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:404)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:304)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```

**Comment by fpj** [comment]
Thanks for reporting this issue, @vedanthh . This is not a pravega issue. This is a deployment issue and if anything an operator issue.

**Comment by sumit-bm** [comment]
Ref: #3687: Seems issue was fixed but again same got reported with 2284 build. Hence this needs to look into.

**Comment by fpj** [comment]
I don't think this is a pravega issue, but since @RaulGracia looked into #3687, I'll defer to him.

**Comment by sumit-bm** [comment]
Thanks @fpj. I put the same tags the ref issue had.

**Comment by fpj** [comment]
Investigating this issue, I found this issue in Pulsar: 

https://github.com/apache/pulsar/issues/3402

which led me to this proposal:

https://github.com/apache/pulsar/wiki/PIP-24%3A-Simplify-memory-settings

which talks about changes to the memory settings. The bookkeeper part of the changes have been implemented in the `4.9` branch:

https://github.com/apache/bookkeeper/blob/branch-4.9/conf/bk_server.conf#L971

but not in the `4.8` branch or earlier. The `4.9` branch is not even stable according to the community, so I recommend that we do not hold `0.5.0` on an upgrade of BK to `4.9+`. It is not even a good idea to upgrade BK without a good amount of testing.

Upgrading BK is already reported in another issue: #2954 . We should raise the version number requested in that issue.

**Comment by RaulGracia** [comment]
As @fpj suggests, this is not a Pravega issue, but a Bookkeeper issue. What I can say so far is the following:
1) The [Pravega Operator](https://github.com/pravega/pravega-operator/blob/3218b6984490fd9204450354d32f7fa355265465/pkg/controller/pravega/bookie.go#L193) sets `-XX:MaxDirectMemorySize=1g`, which is expected to be enough to allocate Netty buffers. If these buffers are not being released properly, maybe Bookkeeper can reach the limit. It is not clear from the logs why this problem arises rarely in our longevity runs, specially if the workload is not very aggressive.
2) Bookies are configured with options like `ExitOnOutOfMemoryError` and `CrashOnOutOfMemoryError` that ensure a fail fast behavior in the case of severe memory problems. As far as I see in the logs, this is not the case and Bookkeeper can proceed normally storing data (in fact, this exception occurs only once in the logs in several days of activity).

@vedanthh @sumit-bm I propose the following action items:
- Close this issue, as this problem is properly handled by Bookkeeper and IO continues working.
- To have a better idea of the memory usage in Bookkeeper under this workload, we reproduce this workload and inspect the CPU and Memory used by the different pods in the system. This may help us to discard (or not) a provisioning problem to be fixed in the operator.

**Comment by fpj** [comment]
we don't necessarily need to close it as @sumit-bm and @vedanthh might want to verify it at some point. my suggestion is that we at least push it to a later release, and associate it to the bk upgrade issue, as it depends on a bk upgrade that is not appropriate for `0.5`.

**Comment by RaulGracia** [comment]
Ok, sounds good @fpj.

**Comment by fpj** [comment]
I take it back, we don't need an upgrade to control that memory allocation, see it here:

https://github.com/apache/bookkeeper/blob/branch-4.7/conf/bk_server.conf#L185

We should increase the value of those three configuration parameters:

```
# The Recv ByteBuf allocator initial buf size.
# byteBufAllocatorSizeInitial=65536

# The Recv ByteBuf allocator min buf size.
# byteBufAllocatorSizeMin=65536

# The Recv ByteBuf allocator max buf size.
# byteBufAllocatorSizeMax=1048576
```

and ion particular of  `byteBufAllocatorSizeMax`. Assuming that there are no leaks, this should do it.

**Comment by andreipaduroiu** [comment]
This has been resolved.
